rules/rulesets.go

```diff
@@ -0,0 +1,236 @@
+	return r.markLiveStatus(ruleSetsTombstoned)
+}
+
+// CanDelete returns true if the ruleset is tombstoned
+// and all the existing rules have expired
+func (r *ruleSets) CanDelete() bool {
+	r.RLock()
+	defer r.RUnlock()
+
+	tombstoned := r.liveStatus == ruleSetsTombstoned
+	expired := len(r.byCutoverAsc) == 0 ||
+		(r.byCutoverAsc[len(r.byCutoverAsc)-1].TombStoned() && !r.byCutoverAsc[len(r.byCutoverAsc)-1].Cutover().After(r.nowFn()))
+	return tombstoned && expired
+}
+
+func (r *ruleSets) markLiveStatus(target ruleSetsStatus) error {
+	r.RLock()
+	if r.liveStatus == target {
+		r.RUnlock()
+		return nil
+	}
+	r.RUnlock()
+
+	r.Lock()
+	defer r.Unlock()
+
+	if r.liveStatus == target {
+		return nil
+	}
+	r.liveStatus = target
+
+	// If the rulesets are resurrected, we start watching updates
+	if target == ruleSetsAlive {
+		return r.Start()
+	}
+
+	// Otherwise the rulesets are deleted, we stop watching updates,
+	// and fetch the latest version in case we have not received
+	// the latest ruleset update from the watch
+	r.Stop()
+	value, err := r.store.Get(r.key)
+	if err != nil {
+		return err
+	}
+	return r.Update(value)
+}
+
+func (r *ruleSets) toRuleSets(value kv.Value) (interface{}, error) {
+	if value == nil {
+		return nil, errNilValue
+	}
+	r.proto.Reset()
+	if err := value.Unmarshal(r.proto); err != nil {
+		return nil, err
+	}
+	return policy.NewRuleSet(r.proto, r.iterFn)
+}
+
+// process adds a new ruleset with the assumption that
+// the cutover time of newer versions will be no earlier than
+// that of older verions, which is guaranteed by ruleset updating APIs
+func (r *ruleSets) process(value interface{}) {
+	rs := value.(policy.RuleSet)
+
+	r.Lock()
+	defer r.Unlock()
+
+	// Find the first item whose cutover time is no earlier than target
+	target := rs.Cutover()
+	idx := sort.Search(len(r.byCutoverAsc), func(i int) bool {
+		return !r.byCutoverAsc[i].Cutover().Before(target)
+	})
+	if idx == len(r.byCutoverAsc) {
+		r.byCutoverAsc = append(r.byCutoverAsc, rs)
+		return
+	}
+
+	// If the item found has the same cutover as the target,
+	// only update it if the target version is higher
+	if r.byCutoverAsc[idx].Cutover() == target {
+		if r.byCutoverAsc[idx].Version() < rs.Version() {
+			r.byCutoverAsc[idx] = rs
+		}
+		return
+	}
+
+	// Otherwise this must be a new version with a unique cutover time,
+	// so we insert it by right-shifting the slice
+	numRulesets := len(r.byCutoverAsc)
+	r.byCutoverAsc = append(r.byCutoverAsc, nil)
+	copy(r.byCutoverAsc[idx+1:numRulesets+1], r.byCutoverAsc[idx:numRulesets])
+	r.byCutoverAsc[idx] = rs
+}
+
+func (r *ruleSets) expire() {
+	r.RLock()
+	if r.expiring {
+		r.RUnlock()
+		return
+	}
+	r.RUnlock()
+
+	r.Lock()
+	defer r.Unlock()
+
+	if r.expiring {
+		return
+	}
+	now := r.nowFn()
+	if idx := r.findWithLock(now); idx > 0 {
+		n := copy(r.byCutoverAsc[0:], r.byCutoverAsc[idx:])
+		for i := n; i < len(r.byCutoverAsc); i++ {
+			r.byCutoverAsc[i] = nil
```

Comment:
This is because even if the slice is shrunk, we are still holding reference to the entire underlying memory chunk, which means if the elements between `n` and `len(r.byCutoverAsc)` hold reference to some expensive nested objects, the garbage collector would need to spend a large amount of time sweeping them, which is certainly undesirable.